Components/Core/Timer.cpp

```diff
@@ -20,15 +25,55 @@ Timer::Timer()
 	// We make a timer named "Timer" with a callback function that does nothing, Autoreload false, and the default period of 1s.
 	// The timer ID is specified as (void *)this to provide a unique ID for each timer object - however this is not necessary for polling timers.
 	// The timer is created in the dormant state.
-	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, empty_callback);
+	assert (rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, empty_callback));
+	timerState = UNINITIALIZED;
 }
 
 /**
  * @brief Changes this timer object's RTOS timer period, returns true on success, returns false on failure (timer command queue full)
  */
 bool Timer::ChangePeriod(const uint32_t period)
 {
-	if (xTimerChangePeriod(rtTimerHandle, period / portTICK_RATE_MS, DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdTRUE)
+	if (xTimerChangePeriod(rtTimerHandle, period / portTICK_RATE_MS, DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdTRUE) {
+		timerState = COUNTING;
 		return true;
+	}
 	return false;
 }
+
+
+bool Timer::StartTimer()
+{
+	if (xTimerStart(rtTimerHandle, DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdPASS) {
+		timerState = COUNTING;
+		return true;
+	}
+	return false;
+}
+
+bool Timer::StopTimer()
+{
+	if (xTimerStart(rtTimerHandle, DEFAULT_TIMER_COMMAND_WAIT_PERIOD) == pdPASS) {
+		timerState = PAUSED;
+		return true;
+	}
+	return false;
+}
+
+CurrentState Timer::completed(TimerHandle_t xTimer)
+{
+	if ((remainingTime = TICKS_TO_MS(xTimerGetExpiryTime(rtTimerHandle) - xTaskGetTickCount()) == 0)) {
+		return COMPLETE;
+	}
+}
+
+CurrentState Timer::GetState(TimerHandle_t xTimer)
+{
+	if (xTimerIsTimerActive(rtTimerHandle) != pdFALSE)
+		return COUNTING;
+	return (completed(rtTimerHandle));
+	return timerState;
```

Comment:
Inaccessible code. Once you've made 'completed' just a GetRemainingTime() function, do something like if(GetRemainingTime() == 0) return COMPLETE;

Also remember some checks are only needed if you are in certain states, you are storing internal state for a reason.

For example, you don't need to check Time Remaining if you are already in UNITIALIZED, COMPLETE or PAUSED, in fact you (probably) only need to check this when it is COUNTING.

I would probably make this a switch (or a set of elifs). Also, please remember to keep the internal state consistent, it's better to set the internal state variable (timerState) then return that at the end, or at least set it before returning. These timers will be used more than once.